neptune-core/src/application/loops/main_loop.rs

```diff
@@ -1696,16 +1708,31 @@ impl MainLoopHandler {
 
                 // Handle incoming connections from peer
                 Ok((stream, peer_address)) = self.incoming_peer_listener.accept() => {
-                    if !precheck_incoming_connection_is_allowed(self.global_state_lock.cli(), peer_address.ip()) {
+                    let ip = peer_address.ip();
+                    if !precheck_incoming_connection_is_allowed(self.global_state_lock.cli(), ip) {
                         continue;
                     }
 
+                    // Is this IP banned through database entry?
+                    let peer_banned = self.global_state_lock.lock_guard().await.net.peer_databases.peer_standings.get(ip).await.is_some_and(|x| x.is_bad());
+                    if peer_banned {
+                        debug!("Banned peer {ip} attempted incoming connection. Hanging up.");
+                        continue;
+                    }
+
+                    // Bump semaphore counter for incoming connections. Should
+                    // be done after the precheck to prevent unnecessary
+                    // acquisitions.
+                    let permit = incoming_connections_limit.clone().acquire_owned().await?;
+
                     let state = self.global_state_lock.lock_guard().await;
                     let main_to_peer_broadcast_rx_clone: broadcast::Receiver<MainToPeerTask> = self.main_to_peer_broadcast_tx.subscribe();
                     let peer_task_to_main_tx_clone: mpsc::Sender<PeerTaskToMain> = self.peer_task_to_main_tx.clone();
                     let own_handshake_data: HandshakeData = state.get_own_handshakedata();
                     let global_state_lock = self.global_state_lock.clone(); // bump arc refcount.
                     let incoming_peer_task_handle = tokio::task::spawn(async move {
+                        let _permit = permit; // drop releases slot
```

Comment:
This is not dropping the permit. If you want to drop the permit here, I suggest to be explicit: `drop(permit);`
If you want to use arcane language features, use `let _ = permit;`. [See also](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#explanation-15).